### DISCUSSION TITLE
"Checksum doesn't match" error on Apple M1/M2/M3

### DISCUSSION THREAD

**Discussion by muffintop98** [ptrumpis/snap-camera-server#81] [Q&A] [answered]
Running snap camera server on MacOS 14.1
I have applied the patch to my Snap Camera app, however, when trying to use a lens like "My New Twin", the lens just "spins" after I select it.

In the Snap Camera log, there is an error:
"Checksum doesn't match for lens archive: 60928810985", and I noticed the lens also does NOT appear in my /Snap Camera/Lenses directory.

Any ideas on how to fix?
Thank you!

**Comment by ptrumpis** [comment]
Looks like the patch was not applied. Something went wrong.
I would suggest to re-apply.

You can upload your binary to [MD5 File Checksum](https://emn178.github.io/online-tools/md5_checksum.html) and compare it to the values below.

**MD5 checksums of macOS binary**

- Original (Codesigned)
  `8dc456e29478a0cdfaedefac282958e7`
- Original (No Codesign)
  `15ad19c477d5d246358d68a711e29a6e`
- Patched (Codesigned)
  `1ac420d1828a3d754e99793af098f830`
- Patched (No Codesign)
 `e2ed1f2e502617392060270fa6e5e979`

**Reply by muffintop98** [reply]
Thank you for the response. I completely removed Snap Camera, then reinstalled version 1.21 and followed the patch steps again. My MD5 hash matches your third bullet point, Patched (Codesigned). Unfortunately, I still get the same result when I try to download certain lenses.

I can see the GET requests in Docker: 
2023-11-04 17:30:08 192.168.65.1 - - [04/Nov/2023:22:30:08 +0000] "GET /3/Tt35Kv2TfdpilezO0ZHal?bo=EhgaABoAMgF9OgEEQgYIofTZmwZIAlASYAE%3D&uc=18 HTTP/1.1" 200 140359 "-" "SnapCamera/1.21.0.0 (macOS 14.1)" "-"

The result is still the same in the Snap Camera log:
VCam::ActionType::SetActiveLensId QVariant(VCam::LensActivationEvent, )
VCam::ActionType::SetCoreProcessingLens QVariant(VCam::CoreProcessingLens, )
VCam::ActionType::ActiveLensChanged QVariant(VCam::LensInfo, )
VCam::ActionType::SetCoreProcessingOverlay QVariant(QImage, QImage(null))
VCam::ActionType::SetActiveLensId QVariant(VCam::LensActivationEvent, )
VCam::ActionType::ActiveLensChanged QVariant(VCam::LensInfo, )
"Checksum doesn't match for lens archive: 60928810985"
"Checksum doesn't match for lens archive: 60928810985"
<img width="915" alt="Screenshot 2023-11-04 at 5 33 07 PM" src="https://github.com/ptrumpis/snap-camera-server/assets/149904093/0cb85718-0d5f-477b-92d3-954f7ee199f2">

**Reply by ptrumpis** [reply]
Force quit snap camera and/or restart your system.

**Reply by muffintop98** [reply]
Still no luck 😢 
Bummer!

**Reply by ptrumpis** [reply]
Are you sure that you replaced the original binary?
https://www.youtube.com/watch?v=lJhL7bA7GGA&t=87s

**Reply by dbacks2196** [reply]
I have the same issue (macOS Sonoma)! Happened both before and after I imported @ptrumpis ' latest assets from Snap into the Snap Camera app bundle. Patched the binary, removed code signing, removed extended attributes, re-code-signed, and then ran the error script. Server is also up and running. Unfortunately, as far as I can tell, no additional lenses were made usable after updating the assets. Any thoughts, anyone? Log file attached if interested.
[Log.txt](https://github.com/ptrumpis/snap-camera-server/files/13866723/Log.txt)

**Reply by ptrumpis** [reply]
Please upload the file `/Applications/Snap Camera.app/Contents/MacOS/Snap Camera` to
https://emn178.github.io/online-tools/md5_checksum.html

What value do you get?

**Reply by dbacks2196** [reply]
Thanks for replying so quickly! Sorry I am just answering now. I get 1765b2647b5d65d470fb104d4a0afdc8. This is the same value I've gotten every time I've applied the signature patch to the executable, but I don't remember seeing it as one of the values other people have gotten.

**Reply by akh925** [reply]
I am also having this issue. I've gone through setup successfully multiple times using an Intel Mac Pre Sonoma. I switched to an M3 Yesterday. I followed all of the setup steps (Including applying the patch, replacing the binary, and running errorfix.sh). Unfortunately, I'm also seeing the exact behavior mentioned above. When activating a lens, I receive a checksum error in logs, and the lens has an infinite spinning/loading icon in the snap camera app and never activates.

**Reply by akh925** [reply, marked as answer]
I finally got this working. I don't know the exact cause but it definitely seems to be a conflict with Mac OS 14/Apple Silicon

Steps to resolve:

1. Patch Binary
2. Make Binary executable
3. Resign App signature `sudo codesign --force --deep --sign - "/Applications/Snap Camera.app"` (You may need to unsign first, I didn't)
 -NOTE: Make sure your Terminal has full disk access, Patrick walks you through how to do that in the Mac video
4. Right-click the Snap Camera application (not the binary)-> Get Info -> Make sure "Open Using Rosetta" is toggled on

After following the steps above, I no longer get the checksum error and my lenses activate properly.

**Comment by ptrumpis** [comment]
You need to read and follow these steps if you want assistance.
I can't help without the output of `errorfix.sh`

1. Make sure to have a fresh installation without any previous patches applied, uninstall and re-install if necessary
2. Original files with original MD5 checksums can be found here: [Download Snap Camera](https://github.com/ptrumpis/snap-camera-server/discussions/6).
3. Docker Server needs to be up and running, you can verify in your browser if https://studio-app.snapchat.com/ is reachable
4. Patch your binary with [Patch Tool](https://ptrumpis.github.io/snap-camera-signature-patch/), you need to see a green CRC checkmark and no error message during patching.
5. Download and replace the binary after patching (watch the video guide if you are not sure how)
6. Now download and run [macOS errorfix.sh](https://raw.githubusercontent.com/ptrumpis/snap-camera-signature-patch/master/macOS/errorfix.sh) inside your Terminal, you don't have to manually execute any other command.
7. Report the output here

**Reply by akh925** [reply]
I posted a workaround. I think the missing step is enabling "Open Using Rosetta" to emulate the intel architecture.

**Reply by ptrumpis** [reply]
Thank you, would be nice to hear verification from @muffintop98 and @dbacks2196

**Reply by dbacks2196** [reply]
Thanks for the instructions, @ptrumpis! And WOW, @akh925! You rock! I can't believe the solution was so simple. Everything works perfectly under Rosetta. Thank you so much.

@ptrumpis, you've truly done an amazing job helping to keep Snap Camera preserved for everyone via local servers. Do you foresee a way in the future that will allow this solution to run without these limitations with Snap Camera's ARM binary? I'm just wondering how we'll be able to do this once Apple deprecates Rosetta 2.

Also, everything's working, but feel free to let me know if you'd still like Terminal output from your shell script.

**Reply by ptrumpis** [reply]
Thank you for your feedback!

> Also, everything's working, but feel free to let me know if you'd still like Terminal output from your shell script.

If everything works, I don't need it.
(The script now includes an update that attempts to open Snap Camera on ARM systems with Rosetta as suggested)

> Do you foresee a way in the future that will allow this solution to run without these limitations with Snap Camera's ARM binary?

I will take a look into it.

**Reply by dbacks2196** [reply]
Thank YOU! And awesome.
